### PR TITLE
Create internal asProp for defaultTheme-util

### LINKED ISSUE #157: Component 'as' prop not working as expected
'as' prop does not work as expected when used with e.g. Heading component. Component styles are not applied to the rendered component.

Tested with a NoteBox component in suomifi-design-system repo the following way:
<Heading.h3 as='h2'>{title}</Heading.h3>
When built with Gatsby the heading styles are not applied.

blocking #155

### PR DESCRIPTION
**Pull request by J-Kallunki**
Styled-components has as-prop that replaces styled-template literal tags' element/component. Our components use internal components with styled that are not the ones to be replaced. Using defaultTheme-util (will be later refactor at another fix) as-prop is now internally changed to asProp and implementations with that when needed. Link-component supports both public API as and internal API asProp.

closes #157

### CODE REVIEW

File: src/core/theme/utils/defaultTheme.ts [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
-  theme = suomifiTheme,
+  theme,
+  as,
   ...props
-}: Partial<T>): T =>
+}: Partial<T> & { as?: asPropType }): T =>
   ({
     ...props,
-    theme,
+    ...(!!as ? { asProp: as } : {}),
```
Comment (J-Kallunki) [inline src/core/theme/utils/defaultTheme.ts]: Our HTML-components have `as`-prop defined as they use `style` from styled-components that can handle element/component change with it. The type definition for it is traversed with `extended` to our final UI-components.
https://www.styled-components.com/docs/api#as-polymorphic-prop

Our UI-components use this defaultTheme-function to set defaults for theme. I'm implementing here that at the same time that prop named `as` is taken away and give as `asProp` - so that every time our internal implementation uses `styled` it does not change the internal component it has. Then we need to internally make implementation for `asProp` where we like to use it - otherwise it does nothing even when defined (originally `as`).

File: src/core/Heading/Heading.tsx [hunk 2]
```diff
@@ -44,6 +46,7 @@ const StyledHeading = styled(
         [smallScreenClassName]: smallScreen,
       })}
       variant={variant === 'h1hero' ? 'h1' : variant}
+      as={asProp}
```
Comment (J-Kallunki) [inline src/core/Heading/Heading.tsx]: I decided to make the internal `asProp` -> `as` implementation here at `src/core`-level component as it has nothing to do with `src/components` level implementations. There it will work just fine to use `as` (supported by styled-components).

File: src/core/theme/utils/defaultTheme.ts [hunk 3]
```diff
@@ -1,12 +1,21 @@
 import { suomifiTheme, ThemeProp } from '../';
+import { asPropType } from '../../../utils/typescript';
 
+/**
+ * Do defaults for component props
+ * Add default theme if theme not defined,
+ * Remove as-prop and change it to asProp if defined
+ * @param props Component props
+ */
 export const withDefaultTheme = <
```
Comment (J-Kallunki) [inline src/core/theme/utils/defaultTheme.ts]: Typings for the function:
- `T extends { theme: ThemeProp; asProp?: asPropType }` : `T` contains theme and optional asProp
- `Partial<T> & { as?: asPropType }` : given function gets object that has `T` (all optional keys) and optional `as`
- that function returns `T`(containing theme and optional asProp as said earlier)

File: src/core/Heading/Heading.tsx [hunk 4]
```diff
@@ -37,7 +37,7 @@ const StyledHeading = styled(
     smallScreen,
     className,
     variant,
-    asProp,
+    asProp, // as-property is defined internally as asProp and need to be implemented back if used
```
Comment (aappoalander) [inline src/core/Heading/Heading.tsx]: I'd clarify this comment a bit further to explain that asProp needs to be implemented per use case / component if support for that is wanted.

File: src/core/Link/Link.tsx [hunk 5]
```diff
@@ -6,17 +6,20 @@ import {
...
 type LinkVariant = 'default' | 'external';
 export interface LinkProps extends CompLinkProps, ThemeComponent {
   variant?: LinkVariant;
+  asProp?: asPropType;
 }
 
-const StyledLink = styled(({ theme, ...passProps }: LinkProps) => (
-  <CompLink {...passProps} />
+const StyledLink = styled(({ theme, asProp, ...passProps }: LinkProps) => (
+  // as-property is defined internally as asProp and need to be implemented back if used
```
Comment (aappoalander) [inline src/core/Link/Link.tsx]: I'd rephrase this "need to be implemented back if used" where ever used to better explain that asProp always needs custom implementation per component if support for it is desired.
